feat(spec): standardizing fury cross-language serialization specification (#1413)
apache/incubator-fury: docs/protocols/xlang_object_graph_spec.md

# Cross language object graph serialization

Fury xlang serialization is an automatic object serialization framework that supports reference and polymorphism.
Fury will convert an object from/to fury xlang serialization binary format.
Fury has two core concepts for xlang serialization:

- **Fury xlang binary format**
- **Framework implemented in different languages to convert object to/from Fury xlang binary format**

The serialization format is a dynamic binary format. The dynamics and reference/polymorphism support make Fury flexible,
much more easy to use, but
also introduce more complexities compared to static serialization frameworks. So the format will be more complex.

## Type Systems

### Data Types

> - bool: A boolean value (true or false).

Reviewer: Can we unify the capitalization style of the first letter of the first word after the type, and use it in uppercase or lowercase?

Author: thanks, updated

> - byte: An 8-bit signed integer.

Reviewer 2: Should `byte` be unsigned? Or, should `byte` have no relationship with sign?

Author: It's int8 here

Author: Maybe we should use int8 instead

- i16: A 16-bit signed integer.
- i32: A 32-bit signed integer.
- i64: A 64-bit signed integer.
- half-float: A 16-bit floating point number.
- float: A 32-bit floating point number.

> - double: A 64-bit floating point number including NaN and Infinity.

Reviewer 2: About naming:

How about fp16, fp32, fp64 or float16, float32, float64?

Author: I think we can use float16, float32, float64 instead. It's the type in golang, arrow, numpy. 

 fp16, fp32, fp64 are more used in deeplearning

Reviewer 2: If we use `floatXX`, I think we should also use `intXX` instead of `iXX`

Author: Updated already in previous commit for update to `floatXX`

> - string: A text string encoded using Latin1/UTF16/UTF-8 encoding.

Reviewer 2: Is Latin1 still widely used? And how about UTF32?

Author: > Is Latin1 still widely used? And how about UTF-32?

UTF-32 use 4 byte for a char, which will bloat the data a lot. 

Actually most chars can be expressed using Latin1, and there are many langauges such as java/python/javascript support `Latin1/UTF-16` natively, so we add such encoding here. When the language support does support the `Latin1/UTF-16` encoding, we can skip the encoding/decoding cost, and using a memory copy to create a string object.

Languages like rust/golang using utf-8 for string encoding, they can still use a copy to create a string object from the serialized data if the data is encoded using utf-8. But if the peer language is Java, a conversion from latin1/utf-16 to utf8 would be needed. But it's Ok, because if we use utf-8 encoding only, the java would need to encode the latin1/utf16 to utf8 when serialization, the cost didn't go way.

Author: > If you wan to support ISO 8859 variants, you should probably name them with names like ISO-8859-1. Unicode encodings have reduced use of ISO 8859 family.

@pjfanning Java/pytho/nodejs all use latin1, how about name it as `Latin1(ISO-8859-1)`

https://docs.python.org/3/library/codecs.html
![image](https://github.com/apache/incubator-fury/assets/12445254/78dbfb45-b162-4d67-9e84-c505ead0f655)
https://nodejs.org/api/buffer.html
![image](https://github.com/apache/incubator-fury/assets/12445254/b9cb6a50-7376-4263-bbaf-b915731e13df)

- enum: a data type consisting of a set of named values. Rust enum with non-predefined field values are not supported as

>   an enum

Reviewer 2: How about the underlying type of enum? i.e. how the enum is stored? 32bit? 64bit?

Author: it's serialized as an unsigned var int, we detailed the format for it in `Value Format` section

Reviewer 2: Got it.

- list: A sequence of objects.
- set: An unordered set of unique elements.

> - map: A map of key-value pairs.

Reviewer 2: Is there any restriction ion the key type?

Author: Nope, it should be immutable, but this is nit related to the protocol

Reviewer 2: I mean, what kind of types is allowed in key?

Author: list/map/set/array/tensor/arrow are not allowed, others are all allowed

Reviewer 2: I think you can add it to the spec.

Author: updated

- time types:

>     - Duration: an absolute length of time independent of any calendar/timezone, as a count of seconds and

Reviewer 2: ```suggestion
    - Duration: an absolute length of time, independent of any calendar/timezone, as a count of seconds and
```

>     - Duration: an absolute length of time independent of any calendar/timezone, as a count of seconds and
>       fractions of seconds at nanosecond resolution.
>     - Timestamp: a point in time independent of any calendar/timezone, as a count of seconds and fractions of
>       seconds at nanosecond resolution. The count is relative to an epoch at UTC midnight on January 1, 1970.

Reviewer 2: ```suggestion
    - Duration: an absolute length of time, independent of any calendar/timezone, as a count of nanoseconds.
    - Timestamp: a point in time independent of any calendar/timezone, as a count of nanoseconds, relative to January 1st, 1970 at 00:00:00 UTC.
```

> - decimal: exact decimal value represented as an integer value in two's complement.

Reviewer 2: What's kind of decimal? decimal32/decimal64 in IEEE754? or an unlimited-precison decimal?

Author: An unlimited-precison decimal, the precison will  be written in the data.

Author: Arrow support decimal128/decimal256 only. I'm not sure whether fury should add such limit.

Reviewer 2: If we allow then it's same as float64/float32, which is very efficient and take fixed storage.

If we have unlimited-precision decimal, should we also have bigint and unlimited-precision float?

Author: The decimal will be serialized like:
```java
buffer.writeVarUint32(value.scale());
buffer.writeVarUint32(value.precision());
buffer.writeVarUint32(bytes.length);
buffer.writeBytes(value.unscaledValue().toByteArray());
```

Could you elaborate how decimal128/decimal256 save space or provide better performance?
One thing I can see is that we can skip write `buffer.writeVarUint32(bytes.length)`, we can take binary took up 128/256 bytes. but other data are still unavoidable.

And if the unscaledValue is small, take `new BigDecimal(BigInteger.valueOf(100), 200, MathContext.DECIMAL128)` for an example, it took only 6 bytes, but when using decimal128 format, it will take 131 bytes?

> - binary: binary data.

Reviewer 2: What's binary data? An array of bytes?

Author: Yep

- array type: only allow numeric component. Other arrays will be taken as List. The implementation should support the
  interoperability between array and list.

>     - array: multiple dimension array which every subarray can have have different size.

Reviewer 2: Could subarray have different type?

Author: nope, array component can be numeric only, it's final.

If you want to store a polymorphic object in array, it will be encoded as a list type.

Author: for array, we use a memory copy for its serialization.

The memory layout is fixed for array in fury

    - int16_array: one dimension int16 array.
    - int32_array: one dimension int32 array.
    - int64_array: one dimension int64 array.
    - half_float_array: one dimension half_float_16 array.
    - float_array: one dimension float32 array.

>     - double_array: one dimension float64 array.

Reviewer 2: How about bool, byte and enum?

Author: I missed bool array, will add it later. 
byte array is just binary, seems unnecessary to add it again.

enum array should use list type

- sparse tensor: a multidimensional array whose elements are almost all zeros.
- arrow record batch: an arrow [record batch](https://arrow.apache.org/docs/cpp/tables.html#record-batches) object.
- arrow table: an arrow [table](https://arrow.apache.org/docs/cpp/tables.html#tables) object.

### Type disambiguation

Due to differences between type systems of languages, those types can't be mapped one-to-one between languages. When
deserializing, Fury use the target data structure type and the data type in the data jointly to determine how to
deserialize and populate the target data structure. For example:

```java
class Foo {
  int[] intArray;
  Object[] objects;
  List<Object> objectList;
}

class Foo2 {
  int[] intArray;
  List<Object> objects;
  List<Object> objectList;
}
```

`intArray` has `int32_array` type. But both `objects` and `objectList` field in the serialize data have `list` data
type. When deserializing, the implementation will create an `Object` array for `objects`, but create a `ArrayList`
for `objectList` to populate it's elements. And the serialized data of `Foo` can be deserialized into `Foo2` too.

### Type ID

> All internal data types are expressed using unsigned ID `-64~-1`. Users can use `0~32703` for representing their types.

Reviewer: > Users can use `0~32703` for representing their types.

Therefore, we need to indicate the ID range on the interface for users, and strictly check the input parameters internally.

Author: Yes, the ID will be checked

> At runtime, all type ids are added by `64`, represented and encoded as an unsigned int.

Reviewer: signed ID `-64 ~ -1` ?

Author: Good catch, updated

> At runtime, all type ids are added by `64`, represented and encoded as an unsigned int.

Reviewer: Is it possible to use uint16 to represent unsigned int ?

deserialize and populate the target data structure. For example:

```java
class Foo {
  int[] intArray;
  Object[] objects;
  List<Object> objectList;
}

class Foo2 {
  int[] intArray;
  List<Object> objects;
  List<Object> objectList;
}
```

`intArray` has `int32_array` type. But both `objects` and `objectList` field in the serialize data have `list` data
type. When deserializing, the implementation will create an `Object` array for `objects`, but create a `ArrayList`
for `objectList` to populate it's elements. And the serialized data of `Foo` can be deserialized into `Foo2` too.

Users can also provide meta hint for fields of a type, or the type whole. Here is an example in java which use
annotation to provide such information.

```java

@TypeInfo(fieldsNullable = false, trackingRef = false, polymorphic = false)
class Foo {
  @FieldInfo(trackingRef = false)
  int[] intArray;
  @FieldInfo(polymorphic = true)
  Object object;
  @FieldInfo(tagId = 1, nullable = true)
  List<Object> objectList;
}
```

Such information can be provided in other languages too:

- cpp: use macro and template.
- golang: use struct tag.
- python: use typehint.
- rust: use macro.

### Type ID

> All internal data types are expressed using a ID in range `-64~-1`. Users can use `0~32703` for representing their

Reviewer 2: ```suggestion
All internal data types are expressed using an ID in range `-64~-1`. Users can use `0~32703` for representing their
```

Reviewer 2: And I'm a little worry about that maybe 64 is not enough in the future.

Author: If we extend it to 128, user registeded types will be encoded by 2 bytes, which will bloat the data.

I have two solutions:
- Preserve type id range `32000~32703` for extending
- Use `-96~1` for internal ID, but it only give 32 ids for users to encode type in one byte

> types. At runtime, all type ids are added by `64`, and then encoded as an unsigned varint.

Reviewer 2: `unsigned varint`

So data type is different from encoding, right? There is no unsigned type, and also varint type.

Author: @PragmaTwice I added `var_int32/fixed_int32/fixed_int64/var_int64/sli_int64` to type systems, user can use type alias ot annotation to mark such types:
```java
class Foo {
  List<@VarInt64 Integer> list;
}
```
```golang
type varint64 int64

type Foo struct {
  []varint64 list
}
```

Author: Unsigned types not added, since they are not supported in every language, take `int64` in java as an example.

| REP_MUL_LOWER_SPECIAL     | every char is in `a-zA-Z._$`                              | replace every upper case char by `\|` + `lower case`, then use `LOWER_SPECIAL`, use this encoding if it's smaller than Encoding `3` |
| LOWER_UPPER_DIGIT_SPECIAL | every char is in `a-zA-Z._$`                              | use `LOWER_UPPER_DIGIT_SPECIAL` encoding if it's smaller than Encoding `2`                                                          |
| UTF8                      | any utf-8 char                                            | use `UTF-8` encoding                                                                                                                |
| Compression               | any utf-8 char                                            | lossless compression                                                                                                                |

Depending on cases, one can choose encoding `flags + data` jointly, uses 3 bits of first byte for flags and other bytes
for data.

## Value Format

### Basic types

#### Bool

- size: 1 byte
- format: 0 for `false`, 1 for `true`

#### Byte

- size: 1 byte
- format: write as pure byte.

> #### Short

Reviewer 2: How about unsigned short?

Author: I changed it to int16

Reviewer 2: How about unsigned integer types?

Author: Could you elaborate more? You mean group all unsigned types together?

Author: I added format for fixed int32/64 format, could you take a look at again?

- size: 2 byte
- byte order: little endian order

#### Unsigned int

- size: 1~5 byte
- Format: The most significant bit (MSB) in every byte indicates whether to have the next byte. If first bit is set
  i.e. `b & 0x80 == 0x80`, then
  the next byte should be read until the first bit of the next byte is unset.

#### Signed int

- size: 1~5 byte
- Format: First convert the number into positive unsigned int by `(v << 1) ^ (v >> 31)` ZigZag algorithm, then encoding
  it as an unsigned int.

#### Unsigned long

- size: 1~9 byte
- Fury PVL(Progressive Variable-length Long) Encoding:
    - positive long format: first bit in every byte indicates whether to have the next byte. If first bit is set
      i.e. `b & 0x80 == 0x80`, then the next byte should be read until the first bit is unset.

> #### Signed long

Reviewer 2: For performance, do we consider to provide a fixed length encoding rather than varint for integer types?

Author: In java, we have a global config for it:
```java
  public FuryBuilder withIntCompressed(boolean intCompressed) {
    this.compressInt = intCompressed;
    return this;
  }

  /**
   * Use variable length encoding for long. Enabled by default, use {@link LongEncoding#SLI} (Small
   * long as int) for long encoding.
   */
  public FuryBuilder withLongCompressed(boolean longCompressed) {
    return withLongCompressed(longCompressed ? LongEncoding.SLI : LongEncoding.LE_RAW_BYTES);
  }

  /** Use variable length encoding for long. */
  public FuryBuilder withLongCompressed(LongEncoding longEncoding) {
    this.longEncoding = Objects.requireNonNull(longEncoding);
    return this;
  }
```

Fury will base this config to generate different code. But it seems we can't do this for c++/rust. I'm considering whether we can add encoding to the type systems. For example, we can have a `fint32/fint64`too.

- size: 1~9 byte
- Fury SLI(Small long as int) Encoding:
    - If long is in [-1073741824, 1073741823], encode as 4 bytes int: `| little-endian: ((int) value) << 1 |`
    - Otherwise write as 9 bytes: `| 0b1 | little-endian 8 bytes long |`

> - Fury PVL(Progressive Variable-length Long) Encoding:

Reviewer 2: How to decide/indicate between these two encoding?

Author: `SLI` is faster, but if most longs are very small, it will bloat the data a little. generally speaking, if users use long, probably the value is too large to hold as an int.

>     - First convert the number into positive unsigned long by ` (v << 1) ^ (v >> 63)` ZigZag algorithm to reduce cost of

Reviewer 2: We don't use zigzag in signed int?

Author: We used zigzag in signed int too.

`varint` for long is computing-extensive, so we also provided `Fury SLI(Small long as int) Encoding` to reduce this cost

Reviewer 2: I see. Maybe give some notes on this doc.

meta programing. But at that time, we don't know the type schema in other processes, so we can't generate the serializer
code for such inconsistent types. We may need to generate the code which has a loop and compare field name one by one to
decide whether deserialize and assign the field or skip the field value.

One fast way is that we can optimize the string comparison into `jump` instructions:

- Assume current type has `n` fields, peer type has `n1` fields.
- Generate an auto growing `field id` from `0` for every sorted field in current type at the compile time.
- Compare the received type meta with current type, generate same id if the field name is same, otherwise generate an
  auto growing id starting from `n`, cache this meta at runtime.
- Iterate the fields of received type meta, use a `switch` to compare the `field id` to deserialize data
  and `assign/skip` field value. **Continuous** field id will be optimized into `jump` in `switch` block, so it will
  very fast.

Here is an example, suppose process A has a class `Foo` with version 1 defined as `Foo1`, process B has a class `Foo`
with version 2 defined as `Foo2`:

```c++
// class Foo with version 1

> class Foo1 {
>   int32_t v1; // id 0
>   std::string v2; // id 1
> }
> // class Foo with version 2
> class Foo2 {
>   // id 0, but will have id 2 in process A
>   bool v0;
>   // id 1, but will have id 0 in process A
>   int32_t v1;
>   // id 2, but will have id 3 in process A
>   int64_t long_value;
>   // id 3, but will have id 1 in process A
>   std::string v2;
>   // id 4, but will have id 4 in process A
>   std::vector<std::string> list;
> }

Reviewer 2: ```suggestion
class Foo1 {
  int32_t v1; // id 0
  std::string v2; // id 1
};
// class Foo with version 2
class Foo2 {
  // id 0, but will have id 2 in process A
  bool v0;
  // id 1, but will have id 0 in process A
  int32_t v1;
  // id 2, but will have id 3 in process A
  int64_t long_value;
  // id 3, but will have id 1 in process A
  std::string v2;
  // id 4, but will have id 4 in process A
  std::vector<std::string> list;
};
```

```

When process A received serialized `Foo2` from process B, here is how it deserialize the data:

```c++
Foo1 &foo1 = xxx;
std::vector<fury::FieldInfo> &field_infos = type_meta.field_infos;
for (auto &field_info : field_infos) {
  switch (field_info.field_id) {
    case 0:

>       foo1.v1 = buffer.read_varint32();

Reviewer 3: In this sample,  the field v1 in Foo1 assigned ` id 0` and it's type was int32, whereas in Foo2 , `id 0 ` refers to v0 which is a boolean. how did `case0:   foo1.v1 = buffer.read_varint32();` worked correctly  under these conditions?

Author: The id won't be written in the serialized data, the whole type meta will carry the field name, so when deserialization, fury can compare the field name and generate a different id for fields not match

Reviewer 3: Ok, i got it

> Foo1 &foo1 = xxx;
> std::vector<fury::FieldInfo> &field_infos = type_meta.field_infos;
> for (auto &field_info : field_infos) {
>   switch (field_info.field_id) {
>     case 0:
>       foo1.v1 = buffer.read_varint32();
>       break;
>     case 1:
>       foo1.v2 = fury.read_string();
>       break;
>     default:
>       fury.skip_data(field_info);
>   }
> }

Reviewer 2: ```suggestion
Foo1 foo1 = ...;
const std::vector<fury::FieldInfo> &field_infos = type_meta.field_infos;
for (const auto &field_info : field_infos) {
  switch (field_info.field_id) {
    case 0:
      foo1.v1 = buffer.read_varint32();
      break;
    case 1:
      foo1.v2 = fury.read_string();
      break;
    default:
      fury.skip_data(field_info);
  }
}
```

| REF VALUE FLAG      | `0`        | This flag indicates the object is referencable and the first time to serialize.                                                                         |

When reference tracking is disabled globally or for specific types, or for certain types within a particular
context(e.g., a field of a type), only the `NULL` and `NOT_NULL VALUE` flags will be used for reference meta.

For languages which doesn't support reference such as rust, reference tracking must be disabled for correct
deserialization by fury rust implementation.

For languages whose object values are not null by default:

- In rust, Fury takes `Option:None` as a null value
- In c++, Fury takes `std::nullopt` as a null value
- In golang, Fury takes `null interface/pointer` as a null value

If one want to deserialize in languages like `Java/Python/JavaScript`, he should mark the type with all fields
not-null by default, or using schema-evolution mode to carry the not-null fields info in the data.

## Type Meta

> For every type to be serialized, it must be registered with an optional ID first. The registered type will have a

Reviewer 3: If the tag is replaced by type_id, does it mean that the order of registration needs to be consistent?

Author: Yes, users can specify an ID when registering, then the order doesn't have to be consistent. Otherwise, the order must be consistent.

Author: Specifing an ID is not that convenient compared to specify the tag, but it's faster and save space too.

Author: This do need more consideration, we must determine whether or not do this tradeoff

user-provided or an auto-growing unsigned int i.e. `type_id`. The registration can be used for security check and type
identification. The id of user registered type will be added by `64` to make space for Fury internal data types.

Depending on whether meta share mode and registration is enabled for current type, Fury will write type meta
differently.

### Schema consistent

If schema consistent mode is enabled globally or enabled for current type, type meta will be written as a fury unsigned
varint of `type_id`.

### Schema evolution

If schema evolution mode is enabled globally or enabled for current type, type meta will be written as follows:

> - If meta share mode is not enabled, type meta will be written as schema consistent mode. Additionally, field meta such

Reviewer 3: What is the meta layout when meta share mode is not enabled ?

Author: If meta share mode is not enabled:
    - If type meta hasn't been written before, the data will be written as:
      ```
      | unsigned varint: 0b11111111 | type def |
      ```
    - If type meta has been written before, the data will be written as:
      ```
      | unsigned varint: written index << 1 | type def |
      ```

  as field type and name will be written with the field value using a key-value like layout.

> - If meta share mode is enabled, type meta will be written as a meta-share encoded binary if type hasn't been written

Reviewer 3: In `meta share` mode, will the `meta-share encode binary` be written at the offset pointed to by the `meta-share start offset`, or at the start of an object in the binary?

Author: - If type meta hasn't been written before, add `type def` to `MetaContext`, then get
  index of the meta in `MetaContext`, write that index as an unsigned varint.
- The type meta in `MetaContext` will be carried into the serialized data when the type is first serialized in a
  batch. That what's `optional 4 bytes` used for. It records the start offset of serialized type defs which not in
  provided `MetaContext` in current object graph serialization. Type defs
  ```python
  buffer.write_var_uint32(len(writting_type_metas))
  for type_meta in writting_type_metas:
      type_meta.write_type_def(buffer)
  ```

  before, otherwise an unsigned varint id which references to previous written type meta will be written.

> ## Meta share

Reviewer 3: We can explain the `Meta share` here，including its usage and the conditions under which it is used.

Author: I updated the meta spec, could you review it again.

> This mode will forbid streaming writing since it needs to look back for update the start offset after the whole object
> graph
> writing and meta collecting is finished. Only in this way we can ensure deserialization failure doesn't lost shared
> meta.
> Meta streamline will be supported in the future for enclosed meta sharing which doesn't cross multiple serializations
> of different objects.

> For Schema consistent mode, type will be encoded as an enumerated string by full type name. Here we mainly describe

Reviewer 3: In consistent mode, the type metadata is written as a varint of `type_id` .  However, in  `Meta share` mode, the full name of the type is written. I believe that  `Meta share` mode is used to eliminate the overhead of metadata transmission. But why is the full name necessary in  `Meta share` mode?

Author: The full name of the type name won't be written, I forget to update it

### string

Format:

```
| header: size << 2 | 2 bits encoding flags | binary data |
```

- `size + encoding` will be concat as a long and encoded as an unsigned var long. The little 2 bits is used for
  encoding:
  0 for `latin`, 1 for `utf-16`, 2 for `utf-8`.
- encoded string binary data based on encoding: `latin/utf-16/utf-8`.

Which encoding to choose:

- For JDK8: fury detect `latin` at runtime, if string is `latin` string, then use `latin` encoding, otherwise
  use `utf-16`.
- For JDK9+: fury use `coder` in `String` object for encoding, `latin`/`utf-16` will be used for encoding.
- If the string is encoded by `utf-8`, then fury will use `utf-8` to decode the data. But currently fury doesn't enable
  utf-8 encoding by default for java. Cross-language string serialization of fury uses `utf-8` by default.

### list

Format:

```
length(unsigned varint) | elements header | elements data
```

#### elements header

In most cases, all elements are same type and not null, elements header will encode those homogeneous
information to avoid the cost of writing it for every element. Specifically, there are four kinds of information

> which will be encoded by elements header, each use one bit:

Reviewer: Can `length+elements` also be concat as a long and encoded as an unsigned var long like string type?

Author: Good suggestion! It's a good optimization for small collection.

But elements header took 4 bits, only 3 bits in a varint of a byte can be used for encoding length, the max value for it is `7`. Then we will use two bytes for length encoding, and the byte saving with this trick will be ineffective for lsit with size larger than `7`

I don't know the common size for collection, so not sure whether we should add such optimization. Either those bits left empty for future extension, or for this optimization.

@PragmaTwice @theweipeng Do you have any suggestions?

Author: I think we can concate length+elements header together

Reviewer 3: @chaokunyang  If it is larger than 7, will the 3 bits be empty?

Author: Nope, it will be used, but next byte will have 3 bits empty

Author: I added this optimization to the spec since empty list and small list is common. thanks @LiangliangSui for this suggestion.

- If track elements ref, use the first bit `0b1` of the header to flag it.
- If the elements have null, use the second bit `0b10` of the header to flag it. If ref tracking is enabled for this
  element type, this flag is invalid.
- If the element types are not the declared type, use the 3rd bit `0b100` of the header to flag it.
- If the element types are different, use the 4rd bit `0b1000` header to flag it.

By default, all bits are unset, which means all elements won't track ref, all elements are same type, not null and
the actual element is the declared type in the custom type field.

#### elements data

Based on the elements header, the serialization of elements data may skip `ref flag`/`null flag`/`element type info`.

> `CollectionSerializer#write/read` can be taken as an example.

Reviewer: Could we add a hyperlink here?

Author: updated

not-null by default, or using schema-evolution mode to carry the not-null fields info in the data.

## Type Meta

For every type to be serialized, it must be registered with an optional ID first. The registered type will have a
user-provided or an auto-growing unsigned int i.e. `type_id`. The registration can be used for security check and type
identification. The id of user registered type will be added by `64` to make space for Fury internal data types.

Depending on whether meta share mode and registration is enabled for current type, Fury will write type meta
differently.

### Schema consistent

If schema consistent mode is enabled globally or enabled for current type, type meta will be written as a fury unsigned
varint of `type_id`.

### Schema evolution

If schema evolution mode is enabled globally or enabled for current type, type meta will be written as follows:

- If meta share mode is not enabled:
    - If type meta hasn't been written before, the data will be written as:
      ```
      | unsigned varint: 0b11111111 | type def |
      ```
    - If type meta has been written before, the data will be written as:
      ```

>       | unsigned varint: written index << 1 | type def |

Reviewer 3: `type def` is unnecessary when it had been written before.   Perhaps  `| unsigned varint: written index << 1 ` is correct

Author: Good catch, I missed it

      ```

> - If meta share mode is enabled:

Reviewer 3: The only difference I noticed between the two modes is that one mode sends metadata and the other mode does not.
If the share mode is not enabled, can we write the meta data at the `meta start offset` ?  When it is disabled, we leave the `meta start offset` empty. so we can more easily understand the difference between the two modes.

Author: update to :

- Normal mode(meta share not enabled):
    - If type meta hasn't been written before, add `type def`
      to `captured_type_defs`: `captured_type_defs[type def] = map size`
    - Get index of the meta in `captured_type_defs`, write that index as an unsigned varint.
    - The meta will be written as:
      ```
      | unsigned varint: written index |
      ```
    - After finished the serialization of the object graph, fury will start to write `captured_type_defs`:
        - Firstly, set current to `meta start offset` of fury header
        - Then write `captured_type_defs` one by one:
          ```python
          buffer.write_var_uint32(len(writting_type_defs))
          for type_meta in writting_type_defs:
          type_meta.write_type_def(buffer)
          ```
- Meta share mode: the writing steps are same as the normal mode, but `captured_type_defs` will be shared across
  multiple serializations of different objects. For example, suppose we have a batch to serialize:
    ```python
    captured_type_defs = {}
    stream = ...
    # add `Type1` to `captured_type_defs` and write `Type1`
    fury.serialize(stream, [Type1()])
    # add `Type2` to `captured_type_defs` and write `Type2`, `Type1` is written before.
    fury.serialize(stream, [Type1(), Type2()])
    # `Type1` and `Type2` are written before, no need to write meta.
    fury.serialize(stream, [Type1(), Type2()])
    ```

- Streaming mode(streaming mode doesn't support meta share):
    - If type meta hasn't been written before, the data will be written as:
      ```
      | unsigned varint: 0b11111111 | type def |
      ```
    - If type meta has been written before, the data will be written as:
      ```
      | unsigned varint: written index << 1 |
      ```
      `written index` is the id in `captured_type_defs`.
    - With this mode, `meta start offset` can be omitted.

For languages which doesn't support reference such as rust, reference tracking must be disabled for correct
deserialization by fury rust implementation.

For languages whose object values are not null by default:

- In rust, Fury takes `Option:None` as a null value
- In c++, Fury takes `std::nullopt` as a null value
- In golang, Fury takes `null interface/pointer` as a null value

If one want to deserialize in languages like `Java/Python/JavaScript`, he should mark the type with all fields
not-null by default, or using schema-evolution mode to carry the not-null fields info in the data.

## Type Meta

For every type to be serialized, it must be registered with an optional ID first. The registered type will have a
user-provided or an auto-growing unsigned int i.e. `type_id`. The registration can be used for security check and type
identification. The id of user registered type will be added by `64` to make space for Fury internal data types.

Depending on whether meta share mode and registration is enabled for current type, Fury will write type meta
differently.

### Schema consistent

> If schema consistent mode is enabled globally or enabled for current type, type meta will be written as a fury unsigned

Reviewer 3: How to determine whether an object is of `Schema consistent` or `Schema evolution` mode  during deserialization. There are no flags in the `type meta` header to indicate this

Author: I updated the spec:

- If schema consistent mode is enabled globally, type meta will be written as a fury unsigned varint of `type_id`.
  Schema evolution related meta will be ignored.
- If schema evolution mode is enabled globally and current class is configured to use schema consistent mode like struct
  vs table in flatbuffers:
    - Type meta will be add to `captured_type_defs`: `captured_type_defs[type def stub] = map size` ahead when
      registering type.
    - Get index of the meta in `captured_type_defs`, write that index as `| unsigned varint: index |`.

    - The meta will be written as:
      ```
      | unsigned varint: written index |
      ```
    - After finished the serialization of the object graph, fury will start to write `captured_type_defs`:
        - Firstly, set current to `meta start offset` of fury header
        - Then write `captured_type_defs` one by one:
          ```python
          buffer.write_var_uint32(len(writting_type_defs))
          for type_meta in writting_type_defs:
          type_meta.write_type_def(buffer)
          ```
- Meta share mode: the writing steps are same as the normal mode, but `captured_type_defs` will be shared across
  multiple serializations of different objects. For example, suppose we have a batch to serialize:
    ```python
    captured_type_defs = {}
    stream = ...
    # add `Type1` to `captured_type_defs` and write `Type1`
    fury.serialize(stream, [Type1()])
    # add `Type2` to `captured_type_defs` and write `Type2`, `Type1` is written before.
    fury.serialize(stream, [Type1(), Type2()])
    # `Type1` and `Type2` are written before, no need to write meta.
    fury.serialize(stream, [Type1(), Type2()])
    ```

> - Streaming mode(streaming mode doesn't support meta share):

Reviewer 3: How can we detect which mode is used? I noticed that there is no flag in the header to indicate the mode

Author: This is configured globally when creating fury. All the fury instance across multiple languages must have same config
